src/triangle_inequality_base_kmeans.cpp

```diff
@@ -14,7 +14,10 @@ void TriangleInequalityBaseKmeans::free() {
     OriginalSpaceKmeans::free();
     delete [] s;
     delete [] upper;
-    delete [] lower;
+
+    if (lower)
+        delete [] lower;
```

Comment:
It's not clear to me why you are making these changes to `lower` here (and below). I worry about the effects in other files that rely on this base class.

I think that if `numLowerBounds == 0`, everything should "just work" (i.e. it's okay to allocate and deallocate zero-length array). Further, with this line in particular, if `lower == NULL`, then it's fine to call `delete` on it.

Reply:
Correct! I will update this.

Reply:
I ran a benchmark using the `driver-experiment` under `valgrind` and verified all algorithms produce the correct result (`-DMONITOR_ACCURACY`) with no leaks reported.